File: simparm/twiddler/ConfigEntry.java

// SimParm: Simple and flexible C++ configuration framework
// Copyright (C) 2007 Australian National University
// 
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2 of the License, or (at your option) any later version.
// 
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
// 
// Contact:
// Kevin Pulo
// kevin.pulo@anu.edu.au
// Leonard Huxley Bldg 56
// Australian National University, ACT, 0200, Australia

//package simparm.twiddler;

import java.io.*;
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
import java.util.*;

abstract class ConfigEntry extends ConfigObject 
    implements FocusListener, ActionListener, MouseListener
{
    private String help;
    private boolean invalid;
    private boolean editable;
    private boolean outputOnChange;
    private HashMap<Component,Color> normalBackground =
         new HashMap<Component,Color>();
    private String configSetIdent;
    private int helpID;

    public String getPrefix() { return "set"; }

    public void mousePressed(MouseEvent e) {}
    public void mouseReleased(MouseEvent e) {}
    public void mouseEntered(MouseEvent e) {}
    public void mouseExited(MouseEvent e) {}

    public void mouseClicked(MouseEvent e) {
       if (e.getButton() == MouseEvent.BUTTON3)
          HelpManager.forID( helpID );
    }

    private ConfigObject.SubPanel subpanel = null;
    private JComponent field = null, label = null;
    private boolean fieldMayExpand;

    public void focusGained(FocusEvent e) {} 
    public void focusLost(FocusEvent e) { commitChanges(); } 

    public void actionPerformed(ActionEvent e) { commitChanges(); }

    abstract void commitChanges();

    private void publishFields() {
        SubPanel to_declare;
        if ( !mayDeclare )
            to_declare = null;
        else if ( subpanel != null && subpanel.label == label 
                                   && subpanel.field == field )
            to_declare = subpanel;
        else
            to_declare = new SubPanel(label, field, fieldMayExpand);
        
        if (to_declare != subpanel) {
            SubPanel sp = subpanel;
            subpanel = to_declare;
            parent.subPanelChange( sp, to_declare, this, 0 );
        }
    }

    protected int determine_offset( ConfigObject fromNode, int offset_in_node ) {
        int rv =  super.determine_offset(fromNode, offset_in_node)
            + ( (subpanel != null && subpanel.label != null) ? 1 : 0)
            + ( (subpanel != null && subpanel.field != null) ? 1 : 0);
        return rv;
    }

    protected void declareDescriptedPanels(boolean inserted) {
        mayDeclare = inserted;
        publishFields();
        super.declareDescriptedPanels( inserted );
    }

    protected void setField(JComponent field) { setField(field, false); }
    protected void setField(JComponent field, boolean mayExpand) {
        this.field = field;
        field.addFocusListener(this);
        field.addMouseListener(this);
        updateEnabled();
        updateVisibility();
        publishFields();
    }

    protected void setLabel(JComponent label) {
        if (label != null) {
            label.setFont(label.getFont().deriveFont(Font.PLAIN));
            label.addMouseListener(this);
        }
        this.label = label;
        publishFields();
    }

    public ConfigEntry() {
            setName("");
            setDesc("");
            this.invalid = false;
            this.editable = true;
            this.outputOnChange = false;

            updateVisibility();
            updateEnabled();
    }

    public ConfigEntry(String name, String desc) {
            setName(name);
            setDesc(desc);
            this.invalid = false;
            this.editable = true;
            this.outputOnChange = false;

            updateVisibility();
            updateEnabled();
    }

    public ConfigEntry(String name, String desc, boolean invalid, boolean viewable, boolean editable) {
            setName(name);
            setDesc(desc);
            this.invalid = invalid;
            this.editable = editable;
            this.outputOnChange = false;

            updateVisibility();
            updateEnabled();
    }

    public void setDesc(String desc) {
        String ls = "<html><p>" + desc + ":</p></html>";
        JLabel nlabel = new JLabel(ls, Label.RIGHT);
        setLabel(nlabel);
        super.setDesc(desc);
    }

    public void setHelp(String help) { 
        if (help == null || help.equals("") || help.equals(" ")) {
            if (this.label != null)
                label.setToolTipText(null);
            if (this.field != null)
                field.setToolTipText(null);
        } else {
            int width = 60;
            String orig = new String(help), umgebrochen= new String();
            while (orig.length() > 0) {
                if (orig.length() <= 60) {
                    umgebrochen += orig;
                    orig = "";
                } else {
                    int stelle = orig.lastIndexOf(' ', 60);
                    umgebrochen += orig.substring(0,stelle) + "<br/>";
                    orig = orig.substring(stelle+1);
                }
            }
            umgebrochen = "<html>" + umgebrochen + "</html>";
            if (this.label != null)
                label.setToolTipText(umgebrochen);
            if (this.field != null)
                field.setToolTipText(umgebrochen);
        }
    }

    public void setInvalid(boolean isInvalid) {
        this.invalid = isInvalid;
    }

    public boolean isEditable() { return editable; }

    public void setEditable(boolean isEditable) {
        editable = isEditable;
        updateEnabled();
    }

    public boolean isOutputOnChange() { return outputOnChange; }
    public void setOutputOnChange(boolean doOutputOnChange)
        { this.outputOnChange = doOutputOnChange; }

    protected void updateVisibility() {
        super.updateVisibility();
        if (field != null) field.setVisible( isViewable() );
        if (label != null) label.setVisible( isViewable() );
    }

    protected void updateEnabled() {
        if (field == null) return;

        boolean editable = isEditable();
        LinkedList<Component> myComponents 
            = new LinkedList<Component>();

        myComponents.add( field );
        if (field instanceof Container) {
            Container c = (Container)field;
            Component[] components = c.getComponents();
            for (Component comp : components)
                myComponents.add( comp );
        }

        for (Component comp : myComponents) {
            if (normalBackground.get(comp) == null)
                normalBackground.put(comp, comp.getBackground());
            comp.setEnabled( editable );
            /* This code would change the background color of
                * elements that are disabled. 
            if ( ! editable) {
                comp.setBackground(Color.white);
            } else
                comp.setBackground(normalBackground.get(comp));
            */
        }
    }

    public void setRequiredUserLevel(int requiredLevel) {
        super.setRequiredUserLevel(requiredLevel);
        updateVisibility();
    }

    protected String getStringValue() { return ""; }

    public void doPrint() { doPrint(null); }
    public void doPrint(Acknowledgement ack_to) {
        print( "= " + getStringValue(), ack_to );
    }

    public boolean processCommand(String line, BufferedReader stream)
        throws IOException 
    {
        String[] split = line.split(" ", 2);
        String command = split[0];
        String arg = (split.length == 2) ? split[1] : "";

        if (command.equals("desc"))
            setDesc(arg);
        else if (command.equals("help"))
            setHelp(arg);
        else if (command.equals("invalid"))
            setInvalid( Boolean.valueOf(arg) );
        else if (command.equals("editable"))
            setEditable( Boolean.valueOf(arg) );
        else if (command.equals("outputOnChange"))
            setOutputOnChange( Boolean.valueOf(arg) );
        else if (command.equals("helpID"))
            helpID = Integer.valueOf(arg);
        else
            return super.processCommand(line, stream);

        return true;
    }

    protected boolean isValue(String str) 
        { return str.equals("value") || str.equals("="); }
}

File: simparm/libsimparm/ConfigEntry.cc

// SimParm: Simple and flexible C++ configuration framework
// Copyright (C) 2007 Australian National University
// 
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2 of the License, or (at your option) any later version.
// 
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
// 
// Contact:
// Kevin Pulo
// kevin.pulo@anu.edu.au
// Leonard Huxley Bldg 56
// Australian National University, ACT, 0200, Australia

#include "Entry.hh"
#include "FileEntry.hh"
#include <stdlib.h>
#include <sstream>
#include <fstream>

namespace simparm {

Entry::Entry(string name, string desc)
: Object(name, desc),
  help("help", ""),
  invalid("invalid", false),
  editable("editable", true),
  outputOnChange("outputOnChange", true),
  helpID("helpID", 0)
{
    push_back(help);
    push_back(invalid);
    push_back(editable);
    push_back(outputOnChange);
    push_back(helpID);
}

Entry::Entry(const Entry& from)
:   Node(from),
    Object(from),
    help(from.help),
    invalid(from.invalid),
    editable(from.editable),
    outputOnChange(from.outputOnChange),
    helpID(from.helpID)
{
    push_back(help);
    push_back(invalid);
    push_back(editable);
    push_back(outputOnChange);
    push_back(helpID);
}

Entry::~Entry() 
{}

void formatParagraph(ostream &o, unsigned int left_col, 
                   unsigned int right_col, const string &s) 
{
   unsigned int pos, lookahead = 0;
   unsigned int cur_col = left_col;
   while (lookahead < s.length()) {
      pos = lookahead;
      if (isalpha(s[lookahead]))
         while (lookahead < s.length() && 
                isalpha(s[lookahead])) lookahead++;
      else
         lookahead++;

      if ((lookahead-pos) > 1+(right_col - cur_col)) {
         o << "\n";
         cur_col = 0;
         while (cur_col < left_col) { cur_col++; o << " "; }
      }
      if (cur_col == left_col && isspace(s[pos]))
         /* skip */;
      else
         o << s.substr(pos, lookahead-pos);
      cur_col += lookahead - pos;
   }
   while (cur_col++ <= right_col) o << " ";
}

void Entry::printHelp(ostream &o) const {
   string n = "--" + name().substr(0, std::min<int>(name().length(), 19));
   formatParagraph(o, 0, 20, n);
   o << "  ";
   formatParagraph(o, 23, 79, desc());
   o << "\n";
   if (help() != "") {
      for (int i = 0; i < 23; i++) o << " ";
      formatParagraph(o, 23, 79, help());
      o << "\n";
   }
}

FileEntry::FileEntry(const FileEntry &entry)
: Node(entry),
  StringEntry(entry), out_stream(NULL), in_stream(NULL),
  default_extension(entry.default_extension)
{
    push_back( default_extension );
}
FileEntry::FileEntry(string name, string desc, string value)
: Node(),
  StringEntry(name, desc, value), out_stream(NULL), in_stream(NULL),
  default_extension("extension", "")
{
    push_back( default_extension );
}
FileEntry::~FileEntry() 
{
    close_output_stream();
    close_input_stream();
}

FileEntry &FileEntry::operator=(const string &v)
   { setValue(v); return *this; }

FileEntry::operator bool() const
   { return value() != ""; }

std::istream& FileEntry::get_input_stream() {
    if ( bool(*this) ) {
        if (in_stream != NULL) {
            return *in_stream;
        } else {
            if (value() == "-")
                in_stream = &std::cin;
            else {
                string filename = value();
                openedIStream.reset(
                    new std::ifstream(filename.c_str(), std::ios::in));
                in_stream = openedIStream.get();
            }
            return *in_stream;
        }
    } else 
        throw std::logic_error("Tried to open stream for FileEntry "
                               "that had no filename set.");
}

std::ostream& FileEntry::get_output_stream(bool append)
 
{
    if ( bool(*this) ) {
        if (out_stream != NULL) {
            return *out_stream;
        } else {
            if (value() == "-")
                out_stream = &std::cout;
            else {
                string filename = value();
                openedOStream.reset(
                    new std::ofstream(filename.c_str(), 
                    std::ios::out | ((append) ? std::ios::app : std::ios::out)));
                out_stream = openedOStream.get();
            }
            return *out_stream;
        }
    } else 
        throw std::logic_error("Tried to open stream for FileEntry "
                               "that had no filename set.");
}

void FileEntry::close_output_stream() {
    if (out_stream != NULL) { 
        openedOStream.reset(NULL);
        out_stream = NULL; 
    }
}
void FileEntry::close_input_stream() {
    if (in_stream != NULL) { 
        openedIStream.reset(NULL);
        in_stream = NULL; 
    }
}

template <> const char *typeName<bool>() { return "Bool"; }
template <> const char *typeName<char>() { return "Char"; }
template <> const char *typeName<unsigned char>() 
    { return "UnsignedChar"; }
template <> const char *typeName<short>() { return "Short"; }
template <> const char *typeName<unsigned short>() 
    { return "UnsignedShort"; }
template <> const char *typeName<int>() { return "Int"; }
template <> const char *typeName<unsigned int>() 
    { return "UnsignedInt"; }
template <> const char *typeName<long>() { return "Long"; }
template <> const char *typeName<unsigned long>() 
    { return "UnsignedLong"; }
template <> const char *typeName<long long>() { return "LongLong"; }
template <> const char *typeName<unsigned long long>() 
    { return "UnsignedLongLong"; }
template <> const char *typeName<float>() { return "Float"; }
template <> const char *typeName<double>() { return "Double"; }
template <> const char *typeName<long double>() 
    { return "LongDouble"; }
template <> const char *typeName<string>() { return "String"; }

}

#include "Entry_Impl.hh"
namespace simparm {
    template class EntryType<string>;
    template class EntryType<bool>;
    template class EntryType<unsigned long>;
};

#include "NumericEntry_Impl.hh"

namespace simparm {
    template class NumericEntry<short int>;
    template class NumericEntry<int>;
    template class NumericEntry<long int>;
    template class NumericEntry<unsigned short int>;
    template class NumericEntry<unsigned int>;
    template class NumericEntry<unsigned long int>;
    template class NumericEntry<float>;
    template class NumericEntry<double>;
    template class NumericEntry<long double>;
};
